### PR TITLE
LPS-201817 KB lock - DM style

### PR DESCRIPTION
**Pull request by robertoDiaz**

### CODE REVIEW

**Comment by robertoDiaz** [comment]
ci:test:sf

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java [hunk 1]
```diff
@@ -571,22 +608,31 @@ public KBArticle expireKBArticle(
...
-		return updateStatus(
-			userId, resourcePrimKey, WorkflowConstants.STATUS_EXPIRED,
-			serviceContext);
+			return updateStatus(
+				userId, resourcePrimKey, WorkflowConstants.STATUS_EXPIRED,
+				serviceContext);
+		}
+		finally {
+			unlockKBArticle(resourcePrimKey);
+		}
```
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:635]: This change would imply calling the Unlock method in Saving MVC action command.

**Review by marcogalluzzi — CHANGES_REQUESTED** [review]

File: modules/apps/knowledge-base/knowledge-base-api/src/main/java/com/liferay/knowledge/base/constants/KBArticleConstants.java [hunk 2]
```diff
@@ -19,6 +21,9 @@ public class KBArticleConstants {
 
 	public static final int DEFAULT_VERSION = 1;
 
+	public static final long LOCK_EXPIRATION_TIME = GetterUtil.getLong(
+		PropsUtil.get("lock.expiration.time." + getClassName()));
```
[the remark this answers is not in the document]
Comment (robertoDiaz) [reply modules/apps/knowledge-base/knowledge-base-api/src/main/java/com/liferay/knowledge/base/constants/KBArticleConstants.java:25]: This is due consistency with DL logic. 

In fact, taking into account how java handles de the constants it's the same, since compiler inlines the value in this cases.

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java [hunk 3]
```diff
@@ -403,106 +404,142 @@ public void deleteGroupKBArticles(long groupId) throws PortalException {
...
+			lockKBArticle(PrincipalThreadLocal.getUserId(), resourcePrimKey);
+		}
+
+		try {
+			return kbArticleLocalService.deleteKBArticle(kbArticle);
+		}
+		finally {
+			unlockKBArticle(resourcePrimKey);
+		}
+	}
```
[the remark this answers is not in the document]
Comment (robertoDiaz) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:435]: I'll remove this, you are right

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java [hunk 4]
```diff
@@ -403,106 +404,142 @@ public void deleteGroupKBArticles(long groupId) throws PortalException {
...
+			resourcePrimKey, version);
 
-		deleteKBArticles(
-			kbArticle.getGroupId(), kbArticle.getResourcePrimKey());
+		if (!hasKBArticleLock(userId, resourcePrimKey)) {
+			lockKBArticle(userId, resourcePrimKey);
+		}
 
-		// Resources
+		try {
```
[the remark this answers is not in the document]
Comment (robertoDiaz) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:454]: As explained before and agreed with @adolfopa we are just making this as consistent with DL as possible, and apply it to DL could require a refactor since lock and unlock logic is not surrounding all methods, and we have to analyze how to apply the "withLock" pattern.

Let's see if it passes Brian's review in order to unblock the dependent stories and work in this refactor after, since it is not mandatory.

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (robertoDiaz) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:635]: Why not? 

If you check the DL logic it works this way:

1: When you start the method we check if the current user has a lock:

`hasKBArticleLock(PrincipalThreadLocal.getUserId(), resourcePrimKey)`

If not, a new lock is been created (if possible, since lock logic controls the use cases to create it or not - expired, same user..):

`lockKBArticle(PrincipalThreadLocal.getUserId(), resourcePrimKey);`

So, when you finish the method, we know that this lock belongs to this user and it could be removed safely.

Do yo see any inconvenience in this workflow?

[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:635]: The use case not handled correctly by the current implementation is the following:

* There is a higher level logic calling the KB Article API. This logic wants to lock a KB article to make some modifications on it
* When starting a lock is acquired on the KB Article
* During the task several calls to the `updateKBArticle` are done
* When the task is finished the KB Article is unlocked.

I think this a reasonable way of thinking about the locking mechanism. However, with the proposed implementation the lock is erroneously released after the first update.

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (robertoDiaz) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java]: I've updated the method and remove the try catch in this case

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (robertoDiaz) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:626]: 1. During the task several calls to the updateKBArticle are done

[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:626]: Hi @adolfopa can you please give your opinion on this? Is the use case I describe something to take into account? Or can we just safely ignore it? Thanks

[the remark this answers is not in the document]
Comment (adolfopa) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:626]: For consistency with DM, the behavior from the edit screen should be (roughly):
1. When clicking on cancel, unlock.
2. When publishing/scheduling, unlock.
3. When saving as draft, keep lock active.

For this we need two "update" operations in the KB API: one that doesn't remove the lock, and another one (analogous to DM's updateAndCheckIn()) that updates and removes the lock. This will be called from the UI/action.

**Comment by robertoDiaz** [comment]
ci:test:sf

File: modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/portlet/action/UpdateKBArticleMVCActionCommand.java [hunk 5]
```diff
@@ -332,19 +335,24 @@ else if (cmd.equals(Constants.UPDATE)) {
...
+					resourcePrimKey, title, content, description, sections,
+					sourceURL, displayDate, expirationDate, reviewDate,
+					selectedFileNames, removeFileEntryIds, serviceContext);
+			}
+			else {
+				kbArticle = _kbArticleService.updateAndUnlockKBArticle(
+					resourcePrimKey, title, content, description, sections,
+					sourceURL, displayDate, expirationDate, reviewDate,
+					selectedFileNames, removeFileEntryIds, serviceContext);
+			}
```
Comment (marcogalluzzi) [inline modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/portlet/action/UpdateKBArticleMVCActionCommand.java:349]: Would it be simpler to do always update and just unlock when action is not ACTION_SAVE_DRAFT? This way the method `updateAndUnlockKBArticle` is not needed. I would also avoid the method name `updateAndUnlockKBArticle` because it seems that the regular `update` doesn't do the unlock (it does, not always, but it does)

```java
			kbArticle = _kbArticleService.updateKBArticle(
				resourcePrimKey, title, content, description, sections,
				sourceURL, displayDate, expirationDate, reviewDate,
				selectedFileNames, removeFileEntryIds, serviceContext);
				
			if (workflowAction != WorkflowConstants.ACTION_SAVE_DRAFT) {
				_kbArticleService.unlockKBArticle(resourcePrimKey);
			}
```

**Review by marcogalluzzi — CHANGES_REQUESTED** [review]

File: modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/portlet/action/UpdateKBArticleMVCActionCommand.java (hunk 5 above)
Comment (robertoDiaz) [reply modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/portlet/action/UpdateKBArticleMVCActionCommand.java:349]: please see https://github.com/liferay-lima/liferay-portal/pull/5080#discussion_r1462053434

Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/portlet/action/UpdateKBArticleMVCActionCommand.java:349]: OK, if @adolfopa likes it go ahead

**Review by marcogalluzzi — APPROVED** [review]

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java [hunk 6]
```diff
@@ -1515,121 +1534,122 @@ public KBArticle updateKBArticle(
...
+
+		_indexKBArticle(kbArticle);
+
+		// Workflow
+
+		_startWorkflowInstance(userId, kbArticle, serviceContext);
+
+		if (autoLock && hasKBArticleLock(userId, resourcePrimKey)) {
 			unlockKBArticle(resourcePrimKey);
 		}
```
Comment (marcogalluzzi) [inline modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:1650]: BTW, why did you remove the `finally` ?

Comment (robertoDiaz) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:1650]: I moved this to the other method, but you are right that it could make sense.

I'll restore it and forward the PR

**Comment by robertoDiaz** [comment]
@adolfopa if you don't mind I'll move this forward

**Comment by robertoDiaz** [comment]
ci:forward

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java [hunk 7]
```diff
@@ -1103,6 +1105,19 @@ public void incrementViewCount(
 			increment);
 	}
 
+	public Lock lockKBArticle(long userId, long resourcePrimKey)
```
Comment (adolfopa) [inline modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:1151]: Aren't you missing an `@Override` annotation here?

**Review by adolfopa — CHANGES_REQUESTED** [review]

File: modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java (hunk 7 above)
Comment (robertoDiaz) [reply modules/apps/knowledge-base/knowledge-base-service/src/main/java/com/liferay/knowledge/base/service/impl/KBArticleLocalServiceImpl.java:1151]: Fixing

**Review by adolfopa — APPROVED** [review]

**Comment by robertoDiaz** [comment]
There's a conflict.

Since it's related with the FF creation, I'm going to change it from LPS-195016 to LPD-11003 in order to handle this as expected with the new jira changes

**Comment by robertoDiaz** [comment]
ci:forward

**Comment by robertoDiaz** [comment]
ci:report: 2529880924

**Comment by robertoDiaz** [comment]
ci:help

**Comment by robertoDiaz** [comment]
ci:report:2529880924

**Comment by marcogalluzzi** [comment]
Hi @robertoDiaz having a look the report I think there is at least this related test failing: KBArticleTrashHandlerTest.testTrashAndDeleteWithDraftStatus

**Comment by robertoDiaz** [comment]
> Hi @robertoDiaz having a look the report I think there is at least this related test failing: KBArticleTrashHandlerTest.testTrashAndDeleteWithDraftStatus

Yes I'm already working on this, no worries.

**Comment by robertoDiaz** [comment]
ci:forward

**Comment by robertoDiaz** [comment]
ci:report:2530786492

**Comment by robertoDiaz** [comment]
Errors seem unrelated. Forcing forward

**Comment by robertoDiaz** [comment]
ci:forward:force
